tools/clang/lib/SPIRV/DeclResultIdMapper.cpp

```diff
@@ -1353,6 +1353,40 @@ uint32_t DeclResultIdMapper::createSpirvStageVar(StageVar *stageVar,
     stageVar->setIsSpirvBuiltin();
     return theBuilder.addStageBuiltinVar(type, sc, BuiltIn::SampleId);
   }
+  // According to DXIL spec, the StencilRef SV can only be used by PSOut.
+  case hlsl::Semantic::Kind::StencilRef: {
+    theBuilder.addExtension("SPV_EXT_shader_stencil_export");
+    theBuilder.requireCapability(spv::Capability::StencilExportEXT);
+
+    stageVar->setIsSpirvBuiltin();
+    return theBuilder.addStageBuiltinVar(type, sc, BuiltIn::FragStencilRefEXT);
+  }
+  // According to DXIL spec, the RenderTargetArrayIndex SV can only be used by
+  // VSIn, VSOut, HSCPIn, HSCPOut, DSIn, DSOut, GSVIn, GSOut, PSIn.
+  // According to Vulkan spec, the Layer BuiltIn can only be used in GSOut and
+  // PSIn.
+  case hlsl::Semantic::Kind::RenderTargetArrayIndex: {
+    switch (sigPointKind) {
+    case hlsl::SigPoint::Kind::VSIn:
+    case hlsl::SigPoint::Kind::VSOut:
+    case hlsl::SigPoint::Kind::HSCPIn:
+    case hlsl::SigPoint::Kind::HSCPOut:
+    case hlsl::SigPoint::Kind::PCOut:
+    case hlsl::SigPoint::Kind::DSIn:
+    case hlsl::SigPoint::Kind::DSCPIn:
+    case hlsl::SigPoint::Kind::DSOut:
+    case hlsl::SigPoint::Kind::GSVIn:
+      return theBuilder.addStageIOVar(type, sc, name.str());
+    case hlsl::SigPoint::Kind::GSOut:
+    case hlsl::SigPoint::Kind::PSIn:
+      theBuilder.requireCapability(spv::Capability::Geometry);
```

Comment:
do we need the Geometry capability even if it is PSIn ?

Reply:
For the `Layer` builtin,
> Layer output by a Geometry Execution Model, input to a Fragment Execution Model, for multi-layer framebuffer. See Vulkan or OpenGL API specifications for more detail.

It explicitly requires the `Geometry` capability. I think it makes sense since this thing is used to communicate info from GSOut -> PSIn.

The validator yells at me if I don't add it.

Reply:
I see. Thanks